File: solutions/cpp/allergies/1/allergies.cpp
#include "allergies.h"

namespace allergies {

allergy_test::allergy_test(int score)
    : score_(score), allergies_({{"eggs", 1},
                                 {"peanuts", 2},
                                 {"shellfish", 4},
                                 {"strawberries", 8},
                                 {"tomatoes", 16},
                                 {"chocolate", 32},
                                 {"pollen", 64},
                                 {"cats", 128}}) {}

bool allergy_test::is_allergic_to(std::string_view allergen) const {
  const auto it = allergies_.find(allergen);
  return (it != allergies_.end()) && (score_ & it->second);
}

std::unordered_set<std::string> allergy_test::get_allergies() const {
  std::unordered_set<std::string> allergies;
  for (const auto it : allergies_)
    if (score_ & it.second)
      allergies.insert(std::string(it.first));
  return allergies;
}

} // namespace allergies

File: solutions/swift/allergies/2/Sources/Allergies/Allergies.swift
class Allergies {
  private let allergens = [
    "eggs", "peanuts", "shellfish", "strawberries", "tomatoes", "chocolate", "pollen", "cats",
  ]
  private let allergies: [String]

  init(_ score: Int) {
    allergies = allergens.enumerated().compactMap { score & (1 << $0) > 0 ? $1 : nil }
  }
  func allergicTo(item: String) -> Bool { allergies.contains(item) }
  func list() -> [String] { allergens.filter { allergicTo(item: $0) } }
}
